src/storage/txn/commands/acquire_pessimistic_lock_resumed.rs

```diff
@@ -0,0 +1,211 @@
+    txn::{
+        acquire_pessimistic_lock,
+        commands::{
+            Command, CommandExt, ReleasedLocks, ResponsePolicy, TypedCommand, WriteCommand,
+            WriteContext, WriteResult, WriteResultLockInfo,
+        },
+        Error, Result,
+    },
+    types::{PessimisticLockParameters, PessimisticLockResults},
+    Error as StorageError, PessimisticLockKeyResult, ProcessResult, Result as StorageResult,
+    Snapshot,
+};
+
+#[derive(Debug)]
+pub struct ResumedPessimisticLockItem {
+    pub key: Key,
+    pub should_not_exist: bool,
+    pub params: PessimisticLockParameters,
+    pub lock_wait_token: LockWaitToken,
+}
+
+command! {
+    /// Acquire a Pessimistic lock on the keys.
+    ///
+    /// This can be rolled back with a [`PessimisticRollback`](Command::PessimisticRollback) command.
+    AcquirePessimisticLockResumed:
+        cmd_ty => StorageResult<PessimisticLockResults>,
+        display => "kv::command::acquirepessimisticlockresumed {:?}",
+        (items),
+        content => {
+            items: Vec<ResumedPessimisticLockItem>,
+        }
+}
+
+impl CommandExt for AcquirePessimisticLockResumed {
+    ctx!();
+    tag!(acquire_pessimistic_lock_resumed);
+    request_type!(KvPessimisticLock);
+
+    property!(can_be_pipelined);
+
+    fn write_bytes(&self) -> usize {
+        self.items
+            .iter()
+            .map(|item| item.key.as_encoded().len())
+            .sum()
+    }
+
+    gen_lock!(items: multiple(|x| &x.key));
+}
+
+impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for AcquirePessimisticLockResumed {
+    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+        let mut modifies = vec![];
+        let mut txn = None;
+        let mut reader: Option<SnapshotReader<S>> = None;
+
+        let mut written_rows = 0;
+        let mut res = PessimisticLockResults::with_capacity(self.items.len());
+        let mut encountered_locks = vec![];
+        let need_old_value = context.extra_op == ExtraOp::ReadOldValue;
+        let mut old_values = OldValues::default();
+
+        let mut new_locked_keys = Vec::with_capacity(self.items.len());
+
+        for item in self.items.into_iter() {
+            let ResumedPessimisticLockItem {
+                key,
+                should_not_exist,
+                params,
+                lock_wait_token,
+            } = item;
+
+            // TODO: Refine the code for rebuilding txn state.
+            if txn
+                .as_ref()
+                .map_or(true, |t: &MvccTxn| t.start_ts != params.start_ts)
+            {
+                if let Some(txn) = txn {
+                    if !txn.is_empty() {
+                        modifies.extend(txn.into_modifies());
+                    }
+                }
+                txn = Some(MvccTxn::new(
+                    params.start_ts,
+                    context.concurrency_manager.clone(),
+                ));
```

Comment:
```suggestion
                if let Some(txn) = txn.replace(MvccTxn::new(...)) {
                    if !txn.is_empty() {
                        modifies.extend(txn.into_modifies());
                    }
                }
```

Reply:
`replace` is a bit simpler. Same for the `reader` below.

Reply:
Is `!txn.is_empty()` necessary? IMO the code will look a bit simpler if removing the `if`.